core: add a UUID to every core surface

jcollie opened the pull request:
- Adds a 128bit UUID to every core surface.
- Expose that UUID as an environment variable.
- Add a function to the core app to search for surfaces by UUID.

mitchellh reviewed (changes requested):
Just a change requested on the dependency.

Other than that, I'm fine with this as it is. I think we'll take a much harder look at surface IDs in the future when we also make libghostty aware of hierarchy (windows vs tabs vs splits) but for now this is fine.

mitchellh reviewed (changes requested):
I don't want to vendor the UUID. I just want to make our own UUID type that copy and pastes what we need. I can take this over.

jcollie commented:
I went ahead and implemented our own UUID type.

pluiedev commented:
One downside to an UUID is that you can't fit it inside a glib.Variant by itself (the type system only supports 64-bit integers at most), which means that when you want to pass the surface ID through our GTK actions, you either have to stuff it inside a string (ew, bad) or split it up into two u64s. Neither option sounds enticing, really...

pluiedev commented:
Plus: u128s are also ill-defined at best in C land (just search "C __int128" and the horrors therein), so *every time* you want to refer to a surface via libghostty you have to specify two u64s (or even worse, a string). Why not just use one single u64 in this case?

jcollie commented:
> Yeah we may be able to get away with an incrementing u64. Surface creation can grab a lock to increment and you’d have to have an unfathomable amount of surfaces to ever overflow.

Incrementing a u64 would def be my _least_ preferred option. So many network protocols have had security issues due to easily predictable identifiers that I just automatically reject that as an option. Plus the need for a mutex.

If we aren't going for "real" UUIDs then `std.crypto.random.int(u64)` is the best way to go IMHO.

mitchellh commented:
Yeah sorry saying incrementing was definitely wrong. I mean something closer to Snowflake.

jcollie committed:
core: add 64 bit unique ID to every core surface

- Expose that ID as the environment variable GHOSTTY_SURFACE to
  processes running in Ghostty surfaces.
- Add a function to the core app to search for surfaces by ID.
- ID is randomly generated, it has no other meaning other than as a
  unique identifier for the surface.
